src/v/cloud_roles/refresh_credentials.cc

```diff
@@ -0,0 +1,241 @@
+          switch (err.error_kind) {
+          case api_request_error_kind::failed_abort:
+              vlog(clrl_log.error, "api request failed: {}", err.reason);
+          case api_request_error_kind::failed_retryable:
+              // handle_response has set the sleep duration, we will retry after
+              // a cool-off period
+              vlog(
+                clrl_log.error,
+                "api request failed (retrying after cool-off period): {}",
+                err.reason);
+          }
+          return ss::now();
+      },
+      [this](credentials creds) {
+          _probe.fetch_success();
+          vlog(clrl_log.info, "fetched credentials {}", creds);
+          return _credentials_update(std::move(creds));
+      });
+}
+
+ss::lowres_clock::duration
+cloud_roles::refresh_credentials::impl::calculate_sleep_duration(
+  uint32_t expiry_sec) const {
+    int sleep = std::floor(expiry_sec * sleep_from_expiry_multiplier);
+    return std::chrono::seconds{sleep};
+}
+
+ss::lowres_clock::duration
+cloud_roles::refresh_credentials::impl::calculate_sleep_duration(
+  std::chrono::system_clock::time_point expires_at) const {
+    auto now = std::chrono::system_clock::now();
+    auto diff = std::chrono::duration_cast<std::chrono::seconds>(
+                  expires_at - now)
+                  .count();
+    return calculate_sleep_duration(diff);
+}
+
+void cloud_roles::refresh_credentials::impl::increment_retries() {
+    _retries += 1;
+    auto sleep_ms = _retry_params.backoff_ms * (2 * _retries);
+    vlog(clrl_log.info, "retry after {} ms", sleep_ms);
+    _sleep_duration = sleep_ms;
+}
+
+cloud_roles::api_response_parse_result
+cloud_roles::refresh_credentials::impl::handle_response(
+  cloud_roles::api_response resp) {
+    if (std::holds_alternative<iobuf>(resp)) {
+        try {
+            auto parsed = parse_response(std::move(std::get<iobuf>(resp)));
+            // If we were retrying on some error and got a correct response,
+            // reset the retry counter
+            if (unlikely(_retries)) {
+                reset_retries();
+            }
+            return parsed;
+        } catch (const std::exception& e) {
+            // Parsing error: it may be a temporary issue with the API, or it
+            // may be a permanent problem with our request.
+            increment_retries();
+            return api_response_parse_error{.reason = e.what()};
+        }
+    } else {
+        if (!std::holds_alternative<api_request_error>(resp)) {
+            throw std::runtime_error(
+              fmt_with_ctx(fmt::format, "unexpected response variant"));
+        }
+        auto error = std::get<api_request_error>(resp);
+        if (error.error_kind == api_request_error_kind::failed_retryable) {
+            increment_retries();
+        }
+        return error;
+    }
+}
+
+ss::future<>
+cloud_roles::refresh_credentials::impl::sleep_until_expiry() const {
+    if (_sleep_duration) {
```

Comment:
Yes maybe it should have a safe default and conditional here removed?

Reply:
the sleep duration is derived from the network response. the first time we call `fetch...credentials` there is no sleep duration, so we just go ahead and call the network api. it returns a response containing credentials and their lifetime.

we then derive the sleep duration from this, and set this value. the next time we call `fetch...` - sleep duration is set and we end up sleeping for that time period. 

additionally in case of errors parsing response/network errors etc, the retry mechanism sets this same sleep interval, so that we do not immediately call the network api in a loop.

We need not avoid sleep the first time we call this API, because we need to acquire credentials as soon as possible to avoid blocking ntp archivers from making calls to S3. 

However we could probably be more defensive here, by making sure that we only skip the sleep for the first call, and then once the first call is made we always sleep to either the set value or the default like you mentioned.

Reply:
right, this is an error on my point. in the original version I had a hard stop to the application in these cases, and in retryable error cases I had `increment_retries` which set an increasing sleep. Now that there is no hard stop to the application, all error cases should call `increment_retries` which will increase the sleep like you mentioned.

Thanks for the catch, I will fix it.

Reply:
I have now moved all retry handling to the top level `refresh_credentials::fetch_and_update_credentials` so it is explicit where we are setting retries and increasing the sleep interval.